Added style based changes to the PR # 1725

prasoon2211 opened the pull request:
This PR makes all style based changes suggested in PR #1725 

@tardyp @sa2ajj @Stibbons @benallard  Please have a look.

Updates TODO list:

List of things to correct in this PR:

- [x] Style-based changes:
  * Dictionary intialization
  * Extra newlines
  * Private memebers prepended with `_`
  * Change some method names
  * Minor refactoring
    * Changes to some tests
    * storage_backends.py to be divided into files.
  * Some error handling for graceful exits in case of failure.

- [ ] Making all superclasses use ABC.
  * Capture
  * CaptureBuildTimes
  * StatsStorageBase

- [ ] Add helper methods for quickly capturing data:
  * Capture a specific data for all builders (properties/times).
  * Capture all properties mathing a regex on all builders.
  * Capture all properties mathing a regex on a list of builders.
  * Capture times (start/end/duration) on a list of builders.

- [ ] Changes to the docs:
  * Fix the format (one sentence per line).
  * Document the changes from the above TODOs.
    * Write about ABC usage and helper methods
  * Add a new tutorial/chapter on the usage of this service.

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -124,10 +154,10 @@ def default_callback(start_time):
             callback = default_callback
-        self._time_type = "start-time"
-        CaptureBuildTimes.__init__(self, builder_name, callback)
+        time_type = "start-time"
```
do we need this variable here?

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -142,10 +172,10 @@ def default_callback(end_time):
             callback = default_callback
-        self._time_type = "end-time"
-        CaptureBuildTimes.__init__(self, builder_name, callback)
+        time_type = "end-time"
```
ditto

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -171,10 +201,10 @@ def default_callback(start_time, end_time):
             callback = default_callback
-        self._time_type = "duration"
-        CaptureBuildTimes.__init__(self, builder_name, callback)
+        time_type = "duration"
```
ditto

sa2ajj commented:
I like the improvements, still work to do.

prasoon2211 commented:
@sa2ajj I've corrected all things you asked and added a few more tests.

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -90,27 +101,45 @@ class CaptureBuildTimes(Capture):
+
+    def _err_msg(self, build_data):
+        if self._time_type == "start-time":
+            capture_class = "CaptureBuildStartTime"
```
you can use `self.__class__.__name__` here

sa2ajj commented:
I'll look a bit more in the evening (iin ~12hrs)

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -25,20 +27,26 @@ class Capture(object):
-        self.callback = callback
+        self._callback = callback
         # parent service and buildmaster to be set when StatsService initialized
         self.parent_svcs = []
```
do we expect external users of `parent_svcs`?

prasoon2211 replied:
Yes. It's used in stats service.

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -61,24 +69,27 @@ def default_callback(props, property_name):
+                             " The property %s not found for build number %s on builder %s."
+                             % (self._property_name, msg['number'], self._builder_name))
+            context = self._defaultContext(msg)
+            series_name = self._builder_name + "-" + self._property_name
```
[nit] `'%s-%s' % (self._builder_name, self._property_name)`

prasoon2211 replied:
Ok.

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -185,24 +205,35 @@ class CaptureData(Capture):
-            callback = lambda x: x
+            callback = identity
 
         routingKey = ("stats-yieldMetricsValue", "stats-yield-data")
```
I believe this needs a better documentation.

prasoon2211 replied:
It's on the Todo-list number 4.

sa2ajj replied:
I understand that, however without that documentation it's a bit difficult to understand what happens here.

sa2ajj commented on master/buildbot/statistics/capture.py:
```diff
@@ -185,24 +205,35 @@ class CaptureData(Capture):
-            for svc in self.parent_svcs:
-                yield threads.deferToThread(svc.postStatsValue, post_data, series_name,
-                                            context)
+            series_name = self._builder_name + "-" + self._data_name
```
[nit] ditto

prasoon2211 replied:
Ok.

sa2ajj commented on master/buildbot/statistics/storage_backends/__init__.py:
```diff
@@ -0,0 +1,19 @@
+from buildbot.statistics.storage_backends.base import StatsStorageBase
+from buildbot.statistics.storage_backends.influxdb_client import InfluxStorageService
+
+__all__ = ['StatsStorageBase', 'InfluxStorageService']
```
I am not sure if we need any content here.  I'd say 'no'.

Stibbons replied:
agree. I don't see the use of __all__ in this case

prasoon2211 replied:
So, should I remove everything from here and leave this file empty or just remove the `__all__`? I'm using it so pylint won't complain.

sa2ajj replied:
Just empty file (optionally, leave the copyright comment)

sa2ajj commented on master/buildbot/statistics/storage_backends/base.py:
```diff
@@ -0,0 +1,27 @@
+    """
+
+    def thd_postStatsValue(self, post_data, series_name, context=None):
+        return defer.succeed(None)
```
Any reason why we want to provide default functionality for this method?

Stibbons replied:
let's abc.abstractmethod here (but add a docstring to explain it should return a deferred)

sa2ajj replied:
I believe without new style classes everywhere (Python's) we can't really use abc :(

prasoon2211 replied:
I did add abc here for the next PR: https://github.com/prasoon2211/buildbot/commit/4b436a997649529f021e5312423cbe5a953139c1

So, this can be merged as is and we can apply next PR to correct this.

sa2ajj commented on master/buildbot/statistics/storage_backends/influxdb_client.py:
```diff
@@ -58,17 +47,20 @@ def __init__(self, url, port, user, password, db, captures,
                                      self.password, self.db)
         self.inited = True
```
This is private member, is it not?

prasoon2211 replied:
Yeah. Forgot about this one.

prasoon2211 replied:
Sorry about this but no, inited is used in the tests (which would make it public).

sa2ajj replied:
I do not think this is a really good justification for making it public.

Stibbons replied:
nothing is really private in python... but I agree, self._inited remove the ambiguity of 'can we access from outside' ?

sa2ajj replied:
side note: try to make a member like `__something` :smile:

Stibbons replied:
it is not recommended to play too much with name mangling :) self._inited or self.inited is ok for me

sa2ajj replied:
I agree with the proposed code.

Me suggesting `__something` was regarding "nothing is really private in python..." :smile:

sa2ajj commented:
Tests fail:
```
[FAIL]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/internet/defer.py", line 1045, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/integration/test_customservices.py", line 32, in test_customService
    self.assertEqual(build['steps'][0]['state_string'], 'num reconfig: 1')
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = u"'echo hei'"
b = 'num reconfig: 1'


buildbot.test.integration.test_customservices.CustomServiceMaster.test_customService
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/internet/defer.py", line 1045, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/util/integration.py", line 167, in tearDown
    print >> dump, "FAILED! dumping build db for debug"
exceptions.TypeError: unsupported operand type(s) for >>: 'builtin_function_or_method' and 'instance'

buildbot.test.integration.test_customservices.CustomServiceMaster.test_customService
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x7f8f500f45f0 [56.6692428589s] called=0 cancelled=0 LoopingCall<60>(heartbeat, *(), **{})()>
<DelayedCall 0x7f8f500dfa70 [596.673021793s] called=0 cancelled=0 LoopingCall<600>(Builder.reclaimAllBuilds, *(), **{})()>
<DelayedCall 0x7f8f500dfb00 [1796.67308497s] called=0 cancelled=0 LoopingCall<1800>(Builder.updateBigStatus, *(), **{})()>
<DelayedCall 0x7f8f500d9c68 [3596.66854692s] called=0 cancelled=0 LoopingCall<3600>(DBConnector._doCleanup, *(), **{})()>

buildbot.test.integration.test_customservices.CustomServiceMaster.test_customService
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/internet/defer.py", line 1045, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/util/integration.py", line 106, in setupConfig
    self.setUpDirs(self.basedir)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/util/dirs.py", line 32, in setUpDirs
    shutil.rmtree(dir)
  File "/opt/python/2.7.9/lib/python2.7/shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "/opt/python/2.7.9/lib/python2.7/shutil.py", line 250, in rmtree
    os.remove(fullname)
exceptions.OSError: [Errno 2] No such file or directory: '/home/travis/build/buildbot/buildbot/_trial_temp/basdir/state.sqlite-shm'

buildbot.test.integration.test_integration_template.ShellMaster.test_shell
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/internet/defer.py", line 1043, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/data/buildrequests.py", line 231, in completeBuildRequests
    brdict = yield self.master.db.buildrequests.getBuildRequest(brid)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 192, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/buildrequests.py", line 69, in thd
    res = conn.execute(q)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1405, in execute
    params)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1538, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1646, in _execute_context
    context)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1639, in _execute_context
    context)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 330, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (OperationalError) no such table: buildrequests u'SELECT buildrequests.id, buildrequests.buildsetid, buildrequests.builderid, buildrequests.priority, buildrequests.complete, buildrequests.results, buildrequests.submitted_at, buildrequests.complete_at, buildrequests.waited_for, buildrequest_claims.brid, buildrequest_claims.masterid, buildrequest_claims.claimed_at, sourcestamps.branch, sourcestamps.repository, sourcestamps.codebase, builders.name AS buildername \nFROM buildrequests LEFT OUTER JOIN buildrequest_claims ON buildrequests.id = buildrequest_claims.brid JOIN buildsets ON buildrequests.buildsetid = buildsets.id JOIN buildset_sourcestamps ON buildsets.id = buildset_sourcestamps.buildsetid JOIN sourcestamps ON buildset_sourcestamps.sourcestampid = sourcestamps.id JOIN builders ON buildrequests.builderid = builders.id \nWHERE buildrequests.id = ?' (1,)

buildbot.test.integration.test_mailnotifier.MailMaster.test_notifiy_for_build
-------------------------------------------------------------------------------
Ran 4603 tests in 178.032s

FAILED (skips=206, failures=1, errors=4, successes=4394)
```

prasoon2211 commented:
https://github.com/prasoon2211/buildbot/commit/5165779ada398bc158ef43ebf0f438f501531e8d#commitcomment-12283631

prasoon2211 commented:
It's odd. The test is passing on my system:

```
(sandbox)[prasoon@jimmypage src]$ git status
On branch todo-pr-1
nothing to commit, working directory clean
(sandbox)[prasoon@jimmypage src]$ git log
commit 5165779ada398bc158ef43ebf0f438f501531e8d
Author: Prasoon Shukla <prasoon92.iitr@gmail.com>
Date:   Mon Jul 20 23:51:56 2015 +0530

    Changed 'name' to 'measurement' in InfluxDB client for InfluDB 0.9.1.
    Used self.__class__.__name__ in CaptureBuild* classes.

commit ca49d8eb8e3baa74eec8334a990174994293702d
Author: Prasoon Shukla <prasoon92.iitr@gmail.com>
Date:   Fri Jul 17 18:07:55 2015 +0530

    Added tests for error catching. Fixed style based errors.

commit cf0a8c186210c3ef691c33c74eb07f2740a0b96c
Author: Prasoon Shukla <prasoon92.iitr@gmail.com>
Date:   Fri Jul 17 00:09:05 2015 +0530

    Fixed style based changes

commit b00daacb5887b4cbb3b27bbdbec5dd6c7cedc152
Merge: 20cfb40 fbc65e5
Author: Mikhail Sobolev <mss@mawhrin.net>
Date:   Thu Jul 16 06:02:50 2015 +0300

    Merge pull request #1764 from jvlomax/documentation_update
    
    Add python3 compatability docs

commit 20cfb409ff9d2d2b7c4aa6073b56a9406cec9708
Merge: a3be9cf bef1b4e
Author: Mikhail Sobolev <mss@mawhrin.net>
Date:   Wed Jul 15 21:11:52 2015 +0300

    Merge pull request #1730 from tardyp/authz
    
    Authz

commit a3be9cf3c3cffd16a7f53d5ed9d06c6d44012156
Merge: 8affe66 49baabc
Author: Mikhail Sobolev <mss@mawhrin.net>
Date:   Wed Jul 15 20:43:35 2015 +0300

    Merge pull request #1751 from jaredgrubb/jgrubb-scheduler-list
    
    Schedulers: let codebases be specified as a list
(sandbox)[prasoon@jimmypage src]$ 
(sandbox)[prasoon@jimmypage src]$ trial buildbot.test.integration.test_customservices.CustomServiceMaster.test_customService
/home/prasoon/work/b/sandbox/local/lib/python2.7/site-packages/twisted/internet/endpoints.py:30: DeprecationWarning: twisted.internet.interfaces.IStreamClientEndpointStringParser was deprecated in Twisted 14.0.0: This interface has been superseded by IStreamClientEndpointStringParserWithReactor.
  from twisted.internet.interfaces import (
buildbot.test.integration.test_customservices
  CustomServiceMaster
    test_customService ...                                                 [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 20.602s

PASSED (successes=1)

```

In any case, I am running the full test suite now. I'll see if it fails.

prasoon2211 commented:
The full test suite just finished running and all tests pass. I don't know what's happening!

Also, travis is passing for the faulty commit on my personal travis account:

https://travis-ci.org/prasoon2211/buildbot/builds/71815491

prasoon2211 commented:
@sa2ajj @Stibbons : I think that this can be merged now. Coveralls seems to be having some trouble (My commit from yesterday still hasn't been processed).

Stibbons commented:
+1

sa2ajj commented on master/buildbot/statistics/__init__.py:
```diff
@@ -19,7 +19,7 @@
 from buildbot.statistics.capture import CaptureProperty
 from buildbot.statistics.stats_service import StatsService
-from buildbot.statistics.storage_backends import InfluxStorageService
+from buildbot.statistics.storage_backends.influxdb_client import InfluxStorageService
```
I thought we agreed to get rid of the content of this file, no?

sa2ajj replied:
Ah, no.  It was about another one.
